src/modules/data/audioStream.c

```diff
@@ -24,13 +26,63 @@ AudioStream* lovrAudioStreamInit(AudioStream* stream, Blob* blob, size_t bufferS
   return stream;
 }
 
+AudioStream* lovrAudioStreamInitRaw(AudioStream* stream, int channelCount, int sampleRate, size_t bufferSize, size_t queueLimitInSamples) {
+  stream->bitDepth = 16;
+  stream->channelCount = channelCount;
+  stream->sampleRate = sampleRate;
+  stream->samples = SIZE_MAX;
+  stream->decoder = NULL;
+  stream->bufferSize = stream->channelCount * bufferSize * sizeof(int16_t);
+  stream->buffer = malloc(stream->bufferSize);
+  lovrAssert(stream->buffer, "Out of memory");
+  stream->blob = NULL;
+  arr_init(&stream->queuedRawBuffers);
+  stream->queueLengthInSamples = 0;
+  stream->queueLimitInSamples = queueLimitInSamples;
+  return stream;
+}
+
 void lovrAudioStreamDestroy(void* ref) {
   AudioStream* stream = ref;
-  stb_vorbis_close(stream->decoder);
-  lovrRelease(Blob, stream->blob);
+  if (stream->decoder) {
+    stb_vorbis_close(stream->decoder);
+    lovrRelease(Blob, stream->blob);
+  } else {
+    for (int i = 0; i < stream->queuedRawBuffers.length; i++) {
+      lovrRelease(Blob, stream->queuedRawBuffers.data[i]);
+    }
+    arr_free(&stream->queuedRawBuffers);
+  }
   free(stream->buffer);
 }
 
+static size_t dequeue_raw(AudioStream* stream, int16_t* destination, size_t sampleCount)
+{
+  if (stream->queuedRawBuffers.length == 0) {
+    return 0;
+  }
+  size_t byteCount = sampleCount * sizeof(int16_t);
+  Blob* blob = stream->queuedRawBuffers.data[0];
+  if (blob->size <= byteCount) {
+    size_t blobSize = blob->size;
+    // blob fits in destination in its entirety. Copy over, free it and remove from start of array.
+    memcpy(destination, blob->data, blobSize);
+    lovrRelease(Blob, blob);
+    arr_splice(&stream->queuedRawBuffers, 0, 1);
+    return blobSize / sizeof(int16_t);
+  } else {
+    // blob is too big. copy all that fits, and put remainder in a new blob in its place.
```

Comment:
I'm wondering if it would be possible to store a cursor into the first blob, to avoid additional allocations/copies for this (potentially common) case.

Details: If the blob is too big, we can read as much as we need and advance a global cursor by that amount.  Then, next time, we can start reading at that cursor.  Any time we nom a blob completely the cursor can be reset to zero.

Reply:
I think that makes perfect sense. I'm not happy about this code either, it'll malloc churn like crazy in bad cases. (and it's not prettilly written :P)

I'll add a counter and avoid the memcpy/realloc!